server/master/src/main/java/org/apache/accumulo/master/upgrade/UpgradeCoordinatorImpl.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.accumulo.master.upgrade;
+
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import org.apache.accumulo.server.ServerConstants;
+import org.apache.accumulo.server.ServerContext;
+import org.apache.accumulo.server.ServerUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableMap;
+
+class UpgradeCoordinatorImpl implements UpgradeCoordinator {
+
+  private static Logger log = LoggerFactory.getLogger(UpgradeCoordinatorImpl.class);
+
+  private ServerContext context;
+  private boolean haveUpgradedZooKeeper = false;
+  private boolean startedMetadataUpgrade = false;
+  private int currentVersion;
+  private Map<Integer,Upgrader> upgraders =
+      ImmutableMap.of(ServerConstants.SHORTEN_RFILE_KEYS, new Upgrader8to9());
+
+  UpgradeCoordinatorImpl(int currentVersion, ServerContext ctx) {
+    this.currentVersion = currentVersion;
+    this.context = ctx;
+  }
+
+  private void handleFailure(Exception e) {
+    log.error("FATAL: Error performing upgrade", e);
+    System.exit(1);
+  }
+
+  @Override
+  public synchronized void upgradeZookeeper() {
+    if (haveUpgradedZooKeeper)
+      throw new IllegalStateException("Only expect this method to be called once");
```

Comment:
Preconditions.checkState() might be a cleaner way to express this?

Reply:
It would be shorter, not sure about clearer.  I think this is fine.

Reply:
I go back and forth on this. Either way, we should probably be consistent about this, and either use Preconditions or not, rather than sometimes do and sometimes don't.